### DISCUSSION TITLE
add update-function to PoiPersistenceManager

### DISCUSSION THREAD

**Discussion by Meibes** [mapsforge/mapsforge#1243] [Ideas]
Hello,

the [PoiPersistenceManager](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-poi/src/main/java/org/mapsforge/poi/storage/PoiPersistenceManager.java) offers functions to insert and remove POIs, but is lacking a function to update them. [DbConstants](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-poi/src/main/java/org/mapsforge/poi/storage/DbConstants.java) already has an [Update-Statement](https://github.com/mapsforge/mapsforge/blob/1e7df8b8bd89ef82d0c504338f3e6dab9c22a0bd/mapsforge-poi/src/main/java/org/mapsforge/poi/storage/DbConstants.java#L81) in it. Would it make sense to implement an update-function in `PoiPersistenceManager` or is there a reason why it isn't already implemented? Because the docs say
>All reading and writing operations are done via classes implementing the PoiPersistenceManager interface. This allows adding, removing and changing POIs at any time.

but I couldn't find functions to _change_ the data in the PoiPersistenceManager.

I would like to add such an update function. I would call the function "updatePointOfInterest" and I think one parameter of the type `PointOfInterest` would be enough. The function would only update the data (tags) in the first step, if I need to update the location later I would send a new pull request.

Also, should I implement this function into `PoiPersistenceManager` or only in `AndroidPoiPersistenceManager`, because I only need it in Android?

Thoughts?

greetings

**Comment by devemux86** [comment]
Let's start with the most important: POI can not work with [Android scoped storage](https://developer.android.com/training/data-storage#scoped-storage) that works with streams (and not file paths).

---

The update statement is used in the poi-writer.
Docs may need updating. The workflow is to use the poi-writer to create the poi and then read them in clients.

_Important: any change in Mapsforge API must be implemented in all supported platforms: Android + Desktop._

---

So we could add a simple `updatePointsOfInterest` for the data and _implement it in Android + Desktop_.

For more complicated updates you can work directly with the poi files, using the platform's database API.
`PoiPersistenceManager` can not include all possible database functions for each table in the poi files.

**Reply by Meibes** [reply]
> POI can not work with Android scoped storage that works with streams

check

>The workflow is to use the poi-writer to create the poi and then read them in clients

check

> So we could add a simple updatePointsOfInterest for the data and implement it in Android + Desktop

the easiest solution would to be limit it to one pointOfInterest first, because the already existing statement is `UPDATE poi_data SET data = ? WHERE id = ?;` - so some function like `updatePointOfInterest(id, tags)` would require the smallest changes to the codebase

> For more complicated updates you can work directly with the poi files, using the platform's database API.
PoiPersistenceManager can not include all possible database functions for each table in the poi files.

I can understand that, but because there is already an update-statement I wanted to use it and other users might also benefit from it

**Comment by Meibes** [comment]
while implementing the update function I stumbled upon a few obstacles that need to be discussed:

1. the `AndroidPoiPersistenceManager` always returns a [mutable set](https://github.com/mapsforge/mapsforge/blob/1e7df8b8bd89ef82d0c504338f3e6dab9c22a0bd/mapsforge-poi-android/src/main/java/org/mapsforge/poi/android/storage/AndroidPoiPersistenceManager.java#L172) while the `PointOfInterest` constructor can potentially return an [unmodifiable set](https://github.com/mapsforge/mapsforge/blob/1e7df8b8bd89ef82d0c504338f3e6dab9c22a0bd/mapsforge-poi/src/main/java/org/mapsforge/poi/storage/PointOfInterest.java#L45) - this could lead to errors -> i would suggest making both unmodifiable (`Collections.unmodifiableSet(tags)`)
2. to update the data for pois, my current approach is to iterate over all pois, create duplicates of all poi with different tags and then call the update-function. another way could be to remove the final-modifier of tags in `PointOfInterest` and add a `setTags`-function - what do you think? I don't think this would fit in, but discussing it maybe is a good idea

greetings

**Reply by devemux86** [reply]
To avoid complications, it is best not to load the library with extra workflows and let it serve its purpose:
the writer writes POIs to the database and the readers read them in the clients.

For additional things like updates, etc. it is best to use the database API of each platform.

---

Many users prefer to use SpatiaLite to access the POI databases to use a richer though heavier API.
